### PR TITLE
fix(install): default to GNU host in Cygwin/MSYS/MinGW environments (#4421)

### PR DESCRIPTION
**Pull request by cachebag**
Addresses first mentioned issue in #4421, modifying the install script to default to `x86_64-pc-windows-gnu` under Cygwin/MSYS when no --default-host is provided. 

The second issue requires changes in the compiled rustup-init binary (src/), which is out of scope for now

### CODE REVIEW

File: rustup-init.sh [hunk 1]
```diff
@@ -179,6 +179,21 @@ main() {
...
+        case "$(uname -s)" in 
+            CYGWIN*|MSYS*|MINGW*)
+                _has_default_host_arg=false
+                for arg in "$@"; do 
+                    case "$arg" in 
+                        --default-host*) _has_default_host_arg=true ;;
+                    esac 
+                done 
+                if [ "$_has_default_host_arg" = false ]; then 
+                    set -- "$@" --default-host x86_64-pc-windows-gnu
```
Comment (rami3l) [inline rustup-init.sh]: Should we consider the possibility of [`aarch64-pc-windows-gnullvm`](https://doc.rust-lang.org/nightly/rustc/platform-support/windows-gnullvm.html) as well? It's a tier 2 target for now.

cc @ChrisDenton

**Review by rami3l — COMMENTED** [review]
Thank you very much :)

However, 4221 is not an issue. Maybe you've got the number wrong?

File: rustup-init.sh [hunk 2]
```diff
@@ -179,6 +179,21 @@ main() {
 
         ignore "$_file" "$@" < /dev/tty
     else
+        # Default to GNU toolchain under Cygwin/MSYS/MinGW environments
+        case "$(uname -s)" in
```
Comment (rami3l) [inline rustup-init.sh]: You should move whatever default host detection logic you may have into the `get_architecture()` function, otherwise it might break the `RUSTUP_INIT_SH_PRINT=arch` functionality.

**Comment by cachebag** [comment]
> However, 4221 is not an issue. Maybe you've got the number wrong?

@rami3l 
Thanks for letting me know! Lmk what else may be needed

File: rustup-init.sh (hunk 1 above)
Comment (ChrisDenton) [reply rustup-init.sh]: Hm... for arm we should probably not touch the default for now, at least until support for gnu on the target is more mainstream. The gnullvm target was only fairly recently promoted to having host tooling and iirc requires installing llvm tools (instead of gnu ones). We could revisit this in the future.

**Review by rami3l — COMMENTED** [review]
How is this different from the existing functionality below?

https://github.com/rust-lang/rustup/blob/7365ab767a586c13f4c48e89ef18664c0705b37b/rustup-init.sh#L425-L427

**Comment by cachebag** [comment]
@rami3l 

> How is this different from the existing functionality below?

I presume that it's essentially the safety net to this, no? The idea here is that the issue at hand was detecting `x86_64-pc-windows-msvc` instead of `x86_64-pc-windows-gnu` in Cygwin despite the existing detection. I figured that a blanket override ensures that shell script users on Windows get GNU regardless of detection edge cases.

Do we address/refactor the existing detection instead?

**Comment by rami3l** [comment]
> > How is this different from the existing functionality below?
> 
> I presume that it's essentially the safety net to this, no? The idea here is that the issue at hand was detecting `x86_64-pc-windows-msvc` instead of `x86_64-pc-windows-gnu` in Cygwin despite the existing detection. I figured that a blanket override ensures that shell script users on Windows get GNU regardless of detection edge cases.
> 
> Do we address/refactor the existing detection instead?

Yes, if this PR is made to address https://github.com/rust-lang/rustup/issues/4421#issuecomment-3316574199. I don't have a Cygwin test environment, but the detection seems right already to me, so this PR seems unnecessary in its current state.

cc @ChrisDenton for further verification.

**Comment by cachebag** [comment]
@rami3l
After looking at this with fresher eyes, I see that I shouldn't simply try to force GNU after detection.
When on Windows, the script now sets `_default_host="--default-host $_arch"` and checks if the user already passed `--default-host` (if so, it clears it to respect user preference). I believe this is what @ChrisDenton was [saying here ](https://github.com/rust-lang/rustup/issues/4421#issuecomment-3316574199)and I apologize for misunderstanding.

> Hm, I would have to look into this more before answering that definitively but I think we can address this by the shell script passing --default-host to rustup (assuming it's not set by the user) and not need any bigger changes.

File: rustup-init.sh [hunk 3]
```diff
@@ -80,6 +80,21 @@ main() {
     local _arch="$RETVAL"
     assert_nz "$_arch" "arch"
 
+    local _default_host=""
+    case "$_arch" in
+        *windows*)
+            _default_host="--default-host $_arch"
```
Comment (rami3l) [inline rustup-init.sh]: I think this is okay, but the variable name here is a bit confusing. Maybe this variable can be renamed to `_default_host_override` to emphasize the fact that it's really an override.

Comment (rami3l) [inline rustup-init.sh]: I think another trick can be used to make this simpler: just call this `--default-host=$_arch` so you can avoid `SC2086`. I promise that the host name doesn't contain whitespaces or whatsoever.

File: rustup-init.sh [hunk 4]
```diff
@@ -80,6 +80,21 @@ main() {
...
+            for arg in "$@"; do
```
Comment (rami3l) [inline rustup-init.sh]: This is not necessary either, IIRC according to the GNU coreutils convention, if we pass `--default-host=A --default-host=B`, then only `B` will be recognized. So just putting the `$_default_host_override` before the rest of the args should be fine.

**Review by rami3l — CHANGES_REQUESTED** [review]

**Review by rami3l — COMMENTED** [review]
@cachebag I'm sorry for the long delay, but your current version is much cleaner now, thanks a lot!

File: rustup-init.sh (hunk 3 above)
Comment (rami3l) [reply rustup-init.sh]: @cachebag Did you forget about this one, or was I analyzing this wrong?

Comment (cachebag) [reply rustup-init.sh]: Latest commit should reflect I was just trying to figure out how to quote `_default_host_override` without passing an empty string @rami3l

**Review by rami3l — APPROVED** [review]
@cachebag I think this code looks okay to merge, but just to be sure, has it been tested to work though?

**Comment by cachebag** [comment]
> @cachebag I think this code looks okay to merge, but just to be sure, has it been tested to work though?
@rami3l 

I unfortunately don't run Windows on any devices. I can ping the original author of the issue to try these changes out or quickly install a VM to test this, I think that should work.

**Comment by rami3l** [comment]
@Petr-Lazecky Would you mind testing out this new installation script on Windows to see if the host detection has changed as expected?

cc @ChrisDenton

**Comment by cachebag** [comment]
@rami3l 

Was able to test this on one of my clients' machines. 

image

**Comment by rami3l** [comment]
@cachebag Looks cool! Merging...
